src/files.ml

```diff
@@ -3,6 +3,211 @@
 open Extra
 open Console
 
+(** Logging function for evaluation. *)
+let log_file = new_logger 'f' "file" "file system"
+let log_file = log_file.logger
+
+(** Representation of module paths and related operations. *)
+module Path =
+  struct
+    (** Representation of a module path (roughly, a file path). *)
+    type module_path = string list
+
+    (** Short synonym of [module_path]. *)
+    type t = module_path
+
+    (** [compare] is a standard comparing function for module paths. *)
+    let compare : t -> t -> int = Pervasives.compare
+
+    (** [pp oc mp] prints [mp] to channel [oc]. *)
+    let pp : module_path pp = fun oc mp ->
+      Format.pp_print_string oc (String.concat "." mp)
+  end
+
+(** Functional maps with module paths as keys. *)
+module PathMap = Map.Make(Path)
+
+(** Synonym of [string] for file paths. *)
+type file_path = string
+
+(** Representation of the mapping from module paths to files. *)
+module ModMap :
+  sig
+    (** Module path mapping. *)
+    type t
+
+    (** [empty] is an empty module path mapping. *)
+    val empty : t
+
+    (** Exception raised if an attempt is made to map an already mapped module
+        (including the root). *)
+    exception Already_mapped
+
+    (** [set_root path m] sets the library root of [m] to be [path]. If it was
+        already set, the the exception [Already_mapped] is raised. *)
+    val set_root : file_path -> t -> t
+
+    (** [add mp path m] extends the mapping [m] by associating the module path
+        [mp] to the file path [path]. In the case where [mp] is already mapped
+        in [m], the exception [Already_mapped] is raised. *)
+    val add : Path.t -> file_path -> t -> t
+
+    (** Exception raised if an attempt is made to use the [get] function prior
+        to the root being set (using [set_root]). *)
+    exception Root_not_set
+
+    (** [get mp m] obtains the file path corresponding to the module path [mp]
+        in [m] (with no particular extension). The exception [Root_not_set] is
+        raised if the root of [m] was not previously set with [set_root]. *)
+    val get : Path.t -> t -> file_path
+
+    (** [iter fn m] calls function [fn] on every binding stored in [m]. *)
+    val iter : (Path.t -> file_path -> unit) -> t -> unit
+  end =
+  struct
+    type t = Node of file_path option * t StrMap.t
```

Comment:
The Node constructor seems useless. Why do you use it?

Reply:
It cannot be left out without the `-rectypes` option.

Reply:
And it does not cost anything.

Reply:
Ok but I don't understand what you try to represent and why it is so complicated. An example would be useful here.

Reply:
Well, it's just a tree, with an optional directory path on each node, and with strings (or rather, module path members) on its branches. But I'll document more later anyway since for now this is still a work in progress.

Reply:
It would be nice to add the example of https://github.com/Deducteam/lambdapi/pull/289#issuecomment-558197206 here.